### ISSUE TITLE
`Collection` has a non trivial performance impact

### ISSUE THREAD

**Issue by wSedlacek** [mikro-orm/mikro-orm#2543]
**Describe the bug**
When using `new Collection()` like with `OneToMany()` as described in the [documentation](https://mikro-orm.io/docs/collections/#onetomany-collections) there is a non trivial performance cost even when there are no rows in the many table.

Request with `new Collection()`
```
wrk -t12 -c400 -d30s http://localhost:3333/api/v1/applications
Running 30s test @ http://localhost:3333/api/v1/applications
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   477.08ms  122.01ms   1.46s    96.86%
    Req/Sec    72.13     27.90   151.00     69.68%
  24939 requests in 30.08s, 275.87MB read
  Socket errors: connect 0, read 720, write 31, timeout 0
Requests/sec:    829.07
Transfer/sec:      9.17MB
```

Request with the following lines removed
```ts
  @OneToMany('Screen', (screen: Screen) => screen.application, {
    cascade: [Cascade.ALL],
    strategy: LoadStrategy.JOINED,
  })
  public screens = new Collection<Screen>(this);
```

```
wrk -t12 -c400 -d30s http://localhost:3333/api/v1/applications
Running 30s test @ http://localhost:3333/api/v1/applications
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   365.51ms  131.34ms   1.52s    96.80%
    Req/Sec    95.78     40.93   212.00     63.01%
  33228 requests in 30.08s, 367.56MB read
  Socket errors: connect 0, read 406, write 66, timeout 0
Requests/sec:   1104.56
Transfer/sec:     12.22MB
```

Looking at a bottom up profile of the request we can see that `Collection` and `ArrayCollection` are having a fairly large amount of time being spent on them.
<img width="1356" alt="Screen Shot 2021-12-16 at 10 39 49" src="https://user-images.githubusercontent.com/8206108/146429691-bd81d83d-4b0b-4611-a7c1-0c1fc0792394.png">

Looking at the flame graph we can see where these cost occurs
<img width="628" alt="Screen Shot 2021-12-16 at 10 42 48" src="https://user-images.githubusercontent.com/8206108/146430140-9e16ca30-4b3d-4957-b136-3a1a5cf224d2.png">

We can see what it looks like without that cost too
<img width="427" alt="Screen Shot 2021-12-16 at 10 43 36" src="https://user-images.githubusercontent.com/8206108/146430189-cb064968-559b-4a9c-be41-1158fa944f1b.png">

In this single request example it was from 0.4ms for a find to a 1ms

**To Reproduce**
1. Create an Entity with a one to many relationship 
2. Query the entity
3. The query takes longer simply because `new Collection()` exists

**Expected behavior**
I would expect there to be some performance cost for creating a new class, but this cost seems like more than expected.
I would expect properties like `lazy` to reduce this impact which in my results they don't

**Additional context**
This was tested in a NestJS application on a M1 Max with 10 applications in the applications table and 0 screens in the screen table.

**Versions**

| Dependency | Version |
| - | - |
| node | 14.18.2 |
| typescript | 4.5.3 |
| mikro-orm | 4.5.9 |
| pg | 8.6.0 |

**Comment by B4nan** [comment]
Please try latest dev version, either 4.x or v5, both recieved recent performance improvements.

**Comment by B4nan** [comment]
> Query the entity
> The query takes longer simply because new Collection() exists

If we are talking about loaded entity, the problem is not with constructing collection instance, but without loading its value, or not? 

Ideally provide a reproduction for this, as I am not fully sure what you are comparing. Note that just the fact that you have additional property will have effect on the speed, it does not have to be the constructor itself. 

To me it is not that obvious what you see in those flame graphs. If you have the property and it is populated, there will be obvious cost for hydrating the entities inside, not really for the fact that there is a collection instance. So please describe what you see there.

**Comment by wSedlacek** [comment]
First, Thank you for your quick responses.

> Please try latest dev version, either 4.x or v5, both recieved recent performance improvements.

When testing with `5.0.0-dec.568` I get the following measurements.

With `Collection()`
```
Running 30s test @ http://localhost:3333/api/v1/applications
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   430.48ms  120.35ms   1.49s    96.98%
    Req/Sec    82.41     48.65   212.00     69.81%
  27834 requests in 30.10s, 307.89MB read
  Socket errors: connect 0, read 7, write 1, timeout 0
Requests/sec:    924.85
Transfer/sec:     10.23MB
```

Without `Collection()`
```
Running 30s test @ http://localhost:3333/api/v1/applications
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   339.07ms  107.84ms   1.31s    97.18%
    Req/Sec   102.12     31.94   191.00     68.15%
  35543 requests in 30.09s, 393.16MB read
  Socket errors: connect 0, read 224, write 64, timeout 0
Requests/sec:   1181.42
Transfer/sec:     13.07MB
```

There is an uplift in both results which is great to see, but there still is a noticeable difference when using the collection.

> Ideally provide a reproduction for this, as I am not fully sure what you are comparing.

I will see if I can create a simple application that demonstrates what I am seeing.

**Comment by wSedlacek** [comment]
@B4nan 

Here is a reproduction.
https://github.com/wSedlacek/mikro-orm-collection-perf

I set it up with NestJS for my connivance, but the metrics should apply to any platform.

**Comment by B4nan** [comment]
So all the time you are comparing 1. having the property defined but not loaded and 2. having it loaded? And you feel like that difference you measured should not be there?

Because to me this is expected, I don't see any significant problem. If you populate a relation, there is a hit, don't expect it to be for free. Both the db query and the hydration will take some time. Given your numbers are basically showing something around 25% penalty, it seems quite fine to me if the difference is loading a relation on entity with very few properties.

**Comment by wSedlacek** [comment]
> So all the time you are comparing 1. having the property defined but not loaded and 2. having it loaded? And you feel like that difference you measured should not be there?
> 
> Because to me this is expected, I don't see any significant problem.


No, this expected. My concern is having a property there not loaded verses not having it at all.
In the reproduction, it would be this expectation.

> There should be minimal difference between /publishers and /authors

In the reproduction it is `Requests/sec:   2728.84` vs `Requests/sec:   2488.79`
That is a 240.05 Requests/sec or roughly 10% which seems like too much the property on Author but not populating it.

**Comment by B4nan** [comment]
Looks like it really is the collection constructor, namely the redefinition of internal properties so they are not enumerable.

**Comment by B4nan** [comment]
Numbers on my end (for latest v5 version)

### Before the fix:

```
➜  mikro-orm-collection-perf git:(main) ✗ ./benchmark.sh
Running 30s test @ http://localhost:3000/publishers
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    61.72ms   12.13ms 275.54ms   88.85%
    Req/Sec   325.06    144.95   656.00     64.03%
  116327 requests in 30.06s, 95.85MB read
  Socket errors: connect 157, read 176, write 0, timeout 0
Requests/sec:   3869.50
Transfer/sec:      3.19MB
Running 30s test @ http://localhost:3000/authors
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    74.78ms   10.30ms 163.61ms   83.39%
    Req/Sec   267.48    150.26   626.00     59.50%
  95824 requests in 30.08s, 78.68MB read
  Socket errors: connect 157, read 345, write 0, timeout 0
Requests/sec:   3186.13
Transfer/sec:      2.62MB
Running 30s test @ http://localhost:3000/authors/0/unloaded-books
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    70.78ms   11.05ms 182.76ms   85.99%
    Req/Sec   282.18    138.49   620.00     63.58%
  101290 requests in 30.07s, 22.70MB read
  Socket errors: connect 157, read 208, write 0, timeout 0
Requests/sec:   3367.94
Transfer/sec:    772.92KB
Running 30s test @ http://localhost:3000/authors/0/books
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   175.91ms   18.93ms 337.24ms   76.89%
    Req/Sec   113.78     67.85   313.00     65.71%
  40719 requests in 30.09s, 17.98MB read
  Socket errors: connect 157, read 148, write 0, timeout 0
Requests/sec:   1353.32
Transfer/sec:    611.90KB
```

### After the fix:

```
➜  mikro-orm-collection-perf git:(main) ✗ ./benchmark.sh
Running 30s test @ http://localhost:3000/publishers
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    61.94ms   10.90ms 245.60ms   88.71%
    Req/Sec   323.37    122.66   656.00     64.90%
  115786 requests in 30.07s, 94.96MB read
  Socket errors: connect 157, read 193, write 0, timeout 0
Requests/sec:   3850.64
Transfer/sec:      3.16MB
Running 30s test @ http://localhost:3000/authors
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    62.45ms    8.18ms 110.70ms   88.16%
    Req/Sec   319.89    143.21   656.00     58.83%
  114835 requests in 30.07s, 92.87MB read
  Socket errors: connect 157, read 153, write 0, timeout 0
Requests/sec:   3818.61
Transfer/sec:      3.09MB
Running 30s test @ http://localhost:3000/authors/0/unloaded-books
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    61.84ms   15.26ms 267.26ms   88.86%
    Req/Sec   323.73    134.07   666.00     68.64%
  116188 requests in 30.07s, 26.04MB read
  Socket errors: connect 157, read 78, write 0, timeout 0
Requests/sec:   3864.18
Transfer/sec:      0.87MB
Running 30s test @ http://localhost:3000/authors/0/books
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   161.95ms   26.45ms 501.43ms   91.40%
    Req/Sec   130.67     98.11   333.00     52.17%
  44246 requests in 30.09s, 19.62MB read
  Socket errors: connect 157, read 133, write 0, timeout 0
Requests/sec:   1470.70
Transfer/sec:    667.85KB
```

**Comment by B4nan** [comment]
Fixed via 8a604a69fe684fac7df8b93a3cc88d1ad47a76ff, will also backport to 4.x

**Comment by wSedlacek** [comment]
Fantastic. Thank you!
